addons/binding/org.openhab.binding.smartmeter/src/main/java/org/openhab/binding/smartmeter/internal/MeterDevice.java

```diff
@@ -0,0 +1,254 @@
+        super();
+        this.deviceId = deviceId;
+        this.valueCache = new HashMap<String, MeterValue<?>>();
+        this.valueChangeListeners = new CopyOnWriteArrayList<>();
+        this.printMeterInfo = true;
+        this.connector = createConnector(serialPortManagerSupplier, serialPort, baudrate, baudrateChangeDelay,
+                protocolMode);
+    }
+
+    protected abstract IMeterReaderConnector<T> createConnector(Supplier<SerialPortManager> serialPortManagerSupplier,
+            String serialPort, int baudrate, int baudrateChangeDelay, ProtocolMode protocolMode);
+
+    /**
+     * Gets the configured deviceId.
+     *
+     * @return the id of the SmlDevice from openHAB configuration.
+     */
+    public String getDeviceId() {
+        return deviceId;
+    }
+
+    /**
+     * Returns the specified OBIS value if available.
+     *
+     * @param obis the OBIS code which value should be retrieved.
+     * @return the OBIS value as String if available - otherwise null.
+     */
+    @Nullable
+    public String getValue(String obisId) {
+        MeterValue<?> smlValue = getSmlValue(obisId);
+        if (smlValue != null) {
+            return smlValue.getValue();
+        }
+        return null;
+    }
+
+    /**
+     * Returns the specified OBIS value if available.
+     *
+     * @param obis the OBIS code which value should be retrieved.
+     * @return the OBIS value if available - otherwise null.
+     */
+    @SuppressWarnings("unchecked")
+    @Nullable
+    public <Q extends Quantity<Q>> MeterValue<Q> getSmlValue(String obisId) {
+
+        if (valueCache.containsKey(obisId)) {
+            return (MeterValue<Q>) valueCache.get(obisId);
+        }
+
+        return null;
+    }
+
+    public Collection<String> getObisCodes() {
+        return new ArrayList<>(this.valueCache.keySet());
+    }
+
+    /**
+     * Read values from this device an store them locally against their OBIS code.
+     *
+     * If there is an error in reading, it will be retried {@value #NUMBER_OF_RETRIES} times. The retry will be delayed
+     * by {@code period} seconds.
+     * If its still failing, the connection will be closed and opened again.
+     *
+     * @return
+     *
+     * @throws Exception
+     */
+    public Disposable readValues(ScheduledExecutorService executorService, Duration period) throws Exception {
+
+        try {
+            final int retryDelay = 2;
+            return Flowable.fromPublisher(connector.getMeterValues(initMessage, period, executorService))
+                    .timeout(30, TimeUnit.SECONDS, Schedulers.from(executorService)).doOnSubscribe(sub -> {
+                        logger.info("Opening connection to {}", getDeviceId());
+                        connector.openConnection();
+                    }).doOnError(ex -> {
+                        logger.error("Failed to read: {}. Closing connection and trying again in {} seconds...; {}",
```

Comment:
Please read through https://www.openhab.org/docs/developer/development/guidelines.html#e-logging and adapt all your logging accordingly.
In short: Unless there are any major problems in the overall system setup/configuration/functionality (i.e. unexpected bug situations), you should never use anything else than "debug" log level.
Instead, use the Thing status to report back any connection/communication issues to the user.

Reply:
hmm... It is also reported to the Thing status.
Regarding logging: If a reading fails most of the times than there could be a general problem which needs to be resolved. Therefore I'd need a stacktrace to tackle down the issue. **Would a warn logging level be ok here?**

Reply:
Yes, as I said, if this here is an "unexpected bug situation", a warn/error log with stacktrace for analysis is absolutely fine. If it is something where you would NOT want people to file a bug report, then better only do debug :-)